NorthstarDLL/mods/modsavefiles.cpp

```diff
@@ -0,0 +1,235 @@
+#include "pch.h"
+#include <filesystem>
+#include <sstream>
+#include <fstream>
+#include "squirrel/squirrel.h"
+#include "util/utils.h"
+#include "mods/modmanager.h"
+#include "rapidjson/document.h"
+#include "rapidjson/writer.h"
+#include "rapidjson/stringbuffer.h"
+#include "config/profile.h"
+#include "rapidjson/error/en.h"
+
+fs::path savePath = fs::path(GetNorthstarPrefix()) / "save_data";
+
+bool ContainsNonASCIIChars(std::string str)
+{
+	// we don't allow null characters either, even if they're ASCII characters because idk if people can
+	// use it to circumvent the file extension suffix.
+	return std::any_of(
+		str.begin(), str.end(), [](char c) { return static_cast<unsigned char>(c) > 127 || static_cast<unsigned char>(c) == '\0'; });
+}
+bool CheckFileName(std::string str)
+{
+	// we don't allow null characters either, even if they're ASCII characters because idk if people can
+	// use it to circumvent the file extension suffix.
+	return std::any_of(
+		str.begin(),
+		str.end(),
+		[](char c)
+		{
+			return static_cast<unsigned char>(c) > 127 || static_cast<unsigned char>(c) == '\0' || static_cast<unsigned char>(c) == '\\' ||
+				   static_cast<unsigned char>(c) == '//';
+		});
+}
+
+ADD_SQFUNC("void", NSSaveFile, "string file, string data", "", ScriptContext::CLIENT | ScriptContext::UI | ScriptContext::SERVER)
+{
+	Mod* mod = g_pSquirrel<context>->getcallingmod(sqvm);
+	if (mod == nullptr)
+	{
+		g_pSquirrel<context>->raiseerror(sqvm, "Has to be called from a mod function!");
+		return SQRESULT_ERROR;
+	}
+
+	std::string fileName = g_pSquirrel<context>->getstring(sqvm, 1);
+	if (CheckFileName(fileName))
+	{
+		g_pSquirrel<context>->raiseerror(
+			sqvm,
+			fmt::format("File name invalid ({})! Make sure it has no '\\', '/' or non-ASCII charcters!", fileName, mod->Name).c_str());
+		return SQRESULT_ERROR;
+	}
+
+	std::string content = g_pSquirrel<context>->getstring(sqvm, 2);
+	if (ContainsNonASCIIChars(content))
```

Comment:
Any particular reason as to disallow using UTF8 characters, for instance?

Reply:
Looking further into this, UTF-8 should work. 

I'd rather err on the side of caution, though (can't really be sure if writing an executable/potentially malicious file with UTF-8 is possible)

This is the same reason NUL characters are banned.

Reply:
I was looking for use-cases requiring this: I thought about storing player names, however I don't think EA accounts allow using special characters ([maybe not in all countries?](https://northstar.cool/assets/NorthstarPromoPoster.png))
But in the end you're right, better be safe than sorry! Authorizing non-ASCII chars can easily be done in a PR later.

Reply:
> I don't think EA accounts allow using special characters

Regardless, you should be storing UIDs in that case, not usernames.

Reply:
Quickly unresolving for now cause I wanna get @wolf109909's opinion in regards to NorthstarCN ^^

Reply:
Quiiiiiiiiiiiickly resolving because we have decided there is no real security benefit for only allowing ASCII characters.